### PR TITLE
Unify editor class and reference search

### PR DESCRIPTION
**Pull request by RayKoopa**
Merges the class hierarchy tree and flat documentation reference search in one list.
animation

The hierarchy can be toggled on or off. This setting is stored.
- When off, it is very similar to the previous flat documentation search.
- When on, it resembles the previous class list, but with added member search. Classes required to build the hierarchy but not actually matching any part of the search term when searching for members are grayed out.

Usability changes / improvements
- Case sensitivity can be toggled (setting is stored).
- The first full match is automatically selected.
- Removed now redundant buttons to open the separate class documentation.
- List can be filtered by item type. This setting is not stored to start new searches later on over all types.
- Tooltip on classes now strips unneccessary newlines at start and end.
- Results without an icon get the "ArrowRight" icon to make the text indented logically. Actual icons to differentiate between methods / constants / etc. would be interesting.
- Window size is stored and restored.
- Class names are matched like members, i.e. by subsequences of the search term. Previously, class names only containing the letters in any order of the search term were found.
- Drawing root lines can be toggled with the already existing scene dock settings.
- `Tree` no longer draws buggy lines when hierarchy lines should be rendered while the root item is hidden.

Code related changes
- Search is now only executed when the window is actually visible. It previously executed even at initial filling, for example when the editor was started.
- Deleted `EditorHelpIndex`. `EditorHelpSearch` (now in separate file) handles both.
- Removed `open_class` signal, now handled by  `go_to_help`.
- Removed `EditorNode` signal `request_help_index`, now handled by `request_help_search`.
- Unified `ScriptEditorBase` signals `request_help_index` and `request_help_search` in `request_help` (tell me if you want to keep the name `request_help_search` instead).
- Rewrote so many parts of `EditorHelpSearch::IncrementalSearch` that I also renamed it to `EditorHelpSearch::Runner` as it isn't just an "incremental" search anymore (more a phased search).
- `ClassDB::is_parent_class` now uses `get_parent_class_nocheck` rather than `get_parent_class` internally since it didn't seem to care about empty strings, causing a lot of logged errors when it was used.

### CODE REVIEW

**Comment by djrm** [comment]
looks good, but it seems a bit confusing since in every other place the left panel (or tree) has some relation or dictates the content on the right side.

**Comment by RayKoopa** [comment]
Yeah, thought the same. Open for ideas.

File: editor/editor_help_search.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+#ifndef EDITOR_HELP_SEARCH_H
+#define EDITOR_HELP_SEARCH_H
+
+#include "editor/editor_help.h"
+#include "editor/editor_plugin.h"
+#include "scene/gui/tree.h"
+#include "editor/code_editor.h"
```
Comment (akien-mga) [inline editor/editor_help_search.h]: Sorry, `clang-format` is quite strict on the alphabetical ordering of includes :P
https://travis-ci.org/godotengine/godot/jobs/314530991

**Comment by RayKoopa** [comment]
I know that, I just don't autoformat right away to see if I missed anything. Will be fixed in the next commit when I'm happy about what djrm mentioned.

**Comment by groud** [comment]
To be honest, I find this interface quite bad. As the gif suggests, you will very likely end-up with redundant results on the left and right side.

Maybe I don't understand perfectly how it should word, but instead of that, I would mix them into a single tree, where matching results are shown, whether it is a class, a method, or anything else. Something like that:
screenshot-2017-12-12 godotengine godot
When a method is matching, it is shown as a child (in the tree) of its class. If only the class name matches, no method is displayed. (the class inheritance is still displayed)

**Comment by RayKoopa** [comment]
Yeah, I don't like it much either. I wanted to unify it into one list already and thought about the same. I can try that if it finds acceptance.

**Comment by groud** [comment]
In that case maybe could more option to filter the results ? Like checkboxes or something like that.

**Comment by RayKoopa** [comment]
I edited my previous comment, you mean something like that?
Could also have a checkbox "display in class hierarchy" and if not, it shows a flat list.

**Comment by groud** [comment]
Hum, this takes a lot of space. Maybe a dropdown list to filter the content instead then ?
And a button to switch between a tree view and a list view.
33884767-4c2732f0-df41-11e7-811f-26d14416f3d8

**Comment by RayKoopa** [comment]
Yeah, that's better. I try that.

**Comment by groud** [comment]
Nice, thanks :)

**Comment by letheed** [comment]
I opened an issue #14319 about better search in Search Help a few days ago.

I'd personally prefer filtering just by typing. Since it's a search box, your hands are already on the keyboard. It's fast to just type in "signal" or "method". Checkboxes and menus force me to grab the mouse again.

**Comment by RayKoopa** [comment]
You can tab and configure the filter like in most other applications. I could still add shortcuts. If you want to filter, you have to do more than typing (except someone wants to add keywords).

**Comment by letheed** [comment]
You mean use Tab to select the checkboxes and Space to toggle them?

**Comment by RayKoopa** [comment]
Something like that, it whatever you need to press to typically use the UI keyboard wise. There can still be hotkeys assigned to toggle each filter.

**Comment by groud** [comment]
Well IMHO, no need for obscure keyboard shortcuts for now. In most cases you'll find the result you are looking for by just typing its name.
The mouse will be needed quite rarely.

**Comment by letheed** [comment]
@RayKoopa I'd rather avoid a plethora of ad-hoc new keyboard shortcuts. As for using Tab, that'd almost unusable. Imagine having to disable all the boxes but the Signal one using Tab and space. At this point using the mouse would be less of a sore.

@groud Well when you're looking for a Signal, anything other than a signal showing up is just noise. Same if you're looking for a Class, or a Property. It's not that you won't find it. It's just that the signals are drowned in irrelevant methods and classes.

The problem with the design you're going for, is that you want all of them included by default, but when you don't you usually want just one. With this design, getting from "all included" to "just signals" means unticking every single other box aside from "signals". I think that's bad and tedious.

**Comment by groud** [comment]
Not if you have a drop-down list as what I suggested. You select either "display all", "signals only", "methods only" etc....

**Comment by letheed** [comment]
Yeah the dropdown solution is a bit better, but then you can't have two at a time.

**Comment by groud** [comment]
>  you can't have two at a time.

This is not important, either you know what you are looking for and you select it, or you don't and the "display all" option is ok.

**Comment by letheed** [comment]
Sometimes you don't know if something's going to be a method/getter or a property.

It's all right if you think that deserves to wait for another PR. But I still think that this is not a really great design. It's be *much* more convenient to be able to select types through search terms

**Comment by groud** [comment]
The proposal seems ok to me. Let's just keep the interface simple by now, if more people complain we will improve it. But for now it does not seems really necessary to me.

**Comment by RayKoopa** [comment]
Do we agree on how the results should look like? For example, the search for "set_name" with the hierarchy displayed would
- show all classes having "\*set_name\*" in their own name (unlikely for the example).
- show all classes being part of the hierarchy up to the child class having any "\*set_name\*" member
- show all the matching members (ohrly)
- **not** show other classes which have no matching members or don't include the search term in their class name.

If that's how you'd like it, I have some trouble implementing it wisely.
- I'd require "looking down" the class hierarchy to filter out upper-level nodes, e.g. from base classes into child classes. The current `DocData` only allows looking upwards  via `String ClassDoc::inherits`.
- I'd have to build a tree storing base down to child classes first to logically wander through it. (The current class tree filter doesn't have to bother about this as it filters out only by class names themselves, not members.)
- The inheritance chain should be stored inside `DocData` itself however (as `Vector<ClassDoc> ClassDoc::child_classes` probably) rather than being (re-)created only for the help dialog, and I'm not sure if anyone wants me to add that to `DocData`.

**Comment by groud** [comment]
> Do we agree on how the results should look like?

For me, the logic you describes looks ok.

If we can avoid recording a link to children classes it's better, avoiding redundancy in data structures is better, unless it's for performance reasons. I would just list the matching results then reconstruct the hierachical structure afterwards.

**Comment by RayKoopa** [comment]
That's kind of the problem, the `Tree` only allows me to _add_ new items to the bottom of one parent node, neither insert or anything (I want to ensure to have child classes at the top before class members). The code I'd have to write for that is surely possible, but will be adventurous to understand later on...

**Comment by RayKoopa** [comment]
Looking fine so far. I'll add remembering the selection when toggling the hierarchy (if possible) before I push.
image

**Comment by groud** [comment]
It looks awesome! Maybe adding the function arguments could be good too?

**Comment by RayKoopa** [comment]
Not for this PR. The search results collect the original `*Doc` classes though so this can be added later. Needs more elaboration on how they're going to be searched (including braces, commas, spaces, or just the parameter names? Ignore searching only the parameter types then? should they be shown or hidden with an option?).

**Comment by YeldhamDev** [comment]
@RayKoopa Gorgeous work! But may I request something? As the Scene Dock and its Node Creation dialog have a option to enabled the drawing of relationship lines, could this have such option as well when in tree mode?

**Comment by RayKoopa** [comment]
Didn't know such a thing even exists. I'll add it as a separate setting for now, but it should probably be a editor-global option to show Tree lines.

**Comment by groud** [comment]
You can use the same setting that is used for the scene I think, just retrieve its value. :)

**Comment by RayKoopa** [comment]
Yep, it'll just have a bad category.

**Comment by RayKoopa** [comment]
Question about case sensitivity and what users would expect.

Currently it behaves as previously:
- Search is case insensitive
- Exact matches are only marked if case matches.

Should case sensitivity behave uniquely?
Should it be toggleable with a button in the toolbar and stored?

**Comment by djrm** [comment]
looking good, one question why the arrows?

**Comment by RayKoopa** [comment]
It's stated in the initial comment:
"Results without an icon get the "ArrowRight" icon to make the text indented logically. Actual icons to differentiate between methods / constants / etc. would be interesting."

I'd prefer other icons, but I have no even more basic ones than the icon of the "Node" (a white circle), which is irritating since everything starts looking like being a Node class.

**Comment by djrm** [comment]
i wouldnt put the class icon in members and methods (this will make the classes easier to identify at first sight), and instead of arrow i will put the object icon, for thigs that are objects, for the non objects maybe the arrow is fine, but im not sure, maybe a new icon would be needed

**Comment by RayKoopa** [comment]
I don't know if an item ultimately inherits from Object at the item creation time without wandering up the tree all the time. That would require ugly code and would question if not all items should use the icon of the nearest class which has one.

**Comment by djrm** [comment]
all the objects that do not have an icon are shown with the object icon, thats how currently works. and it seems that you only need to ask to ClassDB if certain class inherits object, and that only happens if the item does not have icon (not that ugly). And no, items should not get the nearest available icon, just the object icon as a placeholder (like everywhere else).

screenshot from 2017-12-16 18-14-22

**Comment by RayKoopa** [comment]
Reasonable enough. Would look like this.
image

**Comment by djrm** [comment]
@RayKoopa, ok dont get mad at me, but i have a final request.

i made some icons for properties, constants and methods, could you try them out.
[members.zip](https://github.com/godotengine/godot/files/1565233/members.zip)

by the way, great work.

**Comment by RayKoopa** [comment]
Excellent! I just reworked the code to make item creation more flexible and use those icons, I had them in mind aswell.

**Comment by djrm** [comment]
yeah the property icon was supposed to be in lowercase but i think thats not really important.

also i made the VS type icons act as general icons, so now variant types will have icons

screenshot from 2017-12-16 19-55-24

and yes i will make the signal and theme icons.

**Comment by RayKoopa** [comment]
That's great, I wondered why I couldn't see those. Do you think flat list looks okay too?

I'd also have several other icon requests throughout the whole editor since they (soon can be added to menues without making them look dumb if some other items don't have an icon)[https://github.com/godotengine/godot/pull/13983]. For this PR however I'd just need a case sensitive toggle button icon.

**Comment by djrm** [comment]
ok, i made the rest of the icons + match case icon. if you have icon request please ask them here https://github.com/godotengine/godot-design

[members.zip](https://github.com/godotengine/godot/files/1565262/members.zip)

also my previous screenshot is from the master branch plus my changes to the icons, so you will have those once an icon update PR gets merged.

**Comment by RayKoopa** [comment]
I've added those, you can see an updated animation in the PR comment.

**Comment by RayKoopa** [comment]
Weird, am I querying `ClassDB` wrong, or why do some classes not get the Object icon even though they inherit from `Object`?
image

```cpp
if (ui_service->has_icon(p_doc->name, "EditorIcons"))
	icon = ui_service->get_icon(p_doc->name, "EditorIcons");
else if (ClassDB::is_parent_class(p_doc->name, "Object"))
	icon = ui_service->get_icon("Object", "EditorIcons");
String tooltip = p_doc->brief_description.strip_edges();
```

**Comment by RayKoopa** [comment]
I guess they should still get the Object icon then?

**Comment by djrm** [comment]
some are not even godot Objects Like the ResourceLoader, the others im not sure

**Comment by RayKoopa** [comment]
According to DocData, they inherit from `Object`, which is also why they're listed in that indentation there (under `Object`).

**Comment by akien-mga** [comment]
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

**Comment by girng** [comment]
i like this a lot, but i don't think an extra click should be needed (to get to help menu)

**Comment by RayKoopa** [comment]
yes?

**Comment by girng** [comment]
thank you

**Comment by aaronfranke** [comment]
Is the behavior of this feature still the same after the discussions? If not, please update the GIF.

**Comment by RayKoopa** [comment]
Yes it is.

**Comment by RayKoopa** [comment]
How am I supposed to ever resolve these conflicts? Why hasn't this been merged before accepting additional (redundant?) work on this?

**Comment by aaronfranke** [comment]
If you are having trouble with Git, just backup your versions, accept all rebase conflicts in favor of Godot, then add your changes back.

**Comment by RayKoopa** [comment]
Yeah except adding them back causes conflicts over conflicts, and I don't know what's been changed over time which I should try to keep / put into my changes.

**Comment by akien-mga** [comment]
> How am I supposed to ever resolve these conflicts? Why hasn't this been merged before accepting additional (redundant?) work on this?

I'm sorry about that, but you made this contribution at an inconvenient time as we were in release freeze for the massive 3.0 release. Since similarly big contributions had piled up on the master branch during December/January, and more kept coming, it took us months to review and merge them while staying on top of new changes, and some PRs from that time are still waiting, like this one.

Newer and potentially redundant work has been merged because it had a scope easier to review and manage. Big PRs are always harder to handle and can end up having conflicts to resolve before they're eventually merged or rejected.

I'm still interested in the UX improvements made in this PR, for the reference.

**Comment by fire** [comment]
https://github.com/fire/godot/tree/unified_help_search Is rebased on top of master, but it's using your search rather than the current search. I hope you find this useful.

**Comment by akien-mga** [comment]
We've discussed this on IRC today and it would be a "go" once rebased on the current master branch:
```
14:18 <Akien> Next one is #14535
14:18 <IssueBot> #14535: Unify editor class and reference search | https://git.io/vbRPq
14:18 <Akien> Also needs a rebase, and I guess we could ask iFire if he wants to do it since OP is busy - but we need a "go" for the proposed feature first
14:19 <reduz> are we ok with this? I personaly like it, but not sure if there is consense on this
14:19 <Akien> There's been a lot of iteration but I think the gif in the OP is the latest status
14:19 <Akien> I think it's pretty cool myself, and a net improvement over the current dialogs
14:20 <Akien> Usage will tell if it's good enough or needs some more usability changes, but it's a good start
14:22 <Akien> Given the reactions on the PR I guess the community likes it
14:22 <reduz> ok
14:23 <reduz> if iFire wants to give it a try, or OP finds time, would be great
14:24 <Groud[m]> I think it's good
14:25 <Groud[m]> the clean and "case sensitive" buttons are weird though
```

**Comment by RayKoopa** [comment]
To refer to what was said in the IRC, I can confirm that the GIF in the OP is up-to-date with the latest changes I commited back then.

**Comment by moiman100** [comment]
> And about the clean button: it would be nice to have a search box dynamically changing the magnifying glass to an X-clear button in case text is entered.

There is a PR for it #20155

**Comment by akien-mga** [comment]
Closing as the PR has conflicts and no update so it can't be merged for the time being. It's a nice feature though and a new rebased PR would be welcome, either from OP or from another contributor extending the work in this PR (which is what the `salvageable` label means).

Thanks for the initial work, I hope it will make it in the master branch eventually.

**Comment by groud** [comment]
Too bad, this was a great improvement. Hope someone will be interested in rebasing it.

**Pull request closed without merging**